src/ledger/LedgerTxnDataSQL.cpp

```diff
@@ -52,62 +55,222 @@ LedgerTxnRoot::Impl::loadData(LedgerKey const& key) const
     return std::make_shared<LedgerEntry const>(std::move(le));
 }
 
-void
-LedgerTxnRoot::Impl::insertOrUpdateData(LedgerEntry const& entry, bool isInsert)
+class BulkUpsertDataOperation : public DatabaseTypeSpecificOperation
 {
-    auto const& data = entry.data.data();
-    std::string actIDStrKey = KeyUtils::toStrKey(data.accountID);
-    std::string const& dataName = data.dataName;
-    std::string dataValue = decoder::encode_b64(data.dataValue);
+    Database& mDB;
+    std::vector<std::string> mAccountIDs;
+    std::vector<std::string> mDataNames;
+    std::vector<std::string> mDataValues;
+    std::vector<int32_t> mLastModifieds;
 
-    std::string sql;
-    if (isInsert)
+    void
+    accumulateEntry(LedgerEntry const& entry)
     {
-        sql = "INSERT INTO accountdata "
-              "(accountid,dataname,datavalue,lastmodified)"
-              " VALUES (:aid,:dn,:dv,:lm)";
+        assert(entry.data.type() == DATA);
+        DataEntry const& data = entry.data.data();
+        mAccountIDs.emplace_back(KeyUtils::toStrKey(data.accountID));
+        mDataNames.emplace_back(decoder::encode_b64(data.dataName));
+        mDataValues.emplace_back(decoder::encode_b64(data.dataValue));
+        mLastModifieds.emplace_back(
+            unsignedToSigned(entry.lastModifiedLedgerSeq));
     }
-    else
+
+  public:
+    BulkUpsertDataOperation(Database& DB,
+                            std::vector<LedgerEntry> const& entries)
+        : mDB(DB)
     {
-        sql = "UPDATE accountdata SET datavalue=:dv,lastmodified=:lm "
-              " WHERE accountid=:aid AND dataname=:dn";
+        for (auto const& e : entries)
+        {
+            accumulateEntry(e);
+        }
     }
 
-    auto prep = mDatabase.getPreparedStatement(sql);
-    auto& st = prep.statement();
-    st.exchange(soci::use(actIDStrKey, "aid"));
-    st.exchange(soci::use(dataName, "dn"));
-    st.exchange(soci::use(dataValue, "dv"));
-    st.exchange(soci::use(entry.lastModifiedLedgerSeq, "lm"));
-    st.define_and_bind();
-    st.execute(true);
-    if (st.get_affected_rows() != 1)
+    BulkUpsertDataOperation(Database& DB,
+                            std::vector<EntryIterator> const& entryIter)
+        : mDB(DB)
     {
-        throw std::runtime_error("could not update SQL");
+        std::vector<LedgerEntry> entries;
```

Comment:
I think you are reserving space in this but not actually using it.

Reply:
Whoops! Dead code residue. Nice catch, fixed.